accepted/future-releases/records/records-feature-specification.md

```diff
@@ -337,13 +347,42 @@ var c = cond ? a : b; // c has type `Record`.
 
 The greatest lower bound of records with different shapes is `Never`.
 
-### Type inference and promotion
+### Type inference
+
+As usual, we define type inference for record expressions with respect to a
+context type schema which is determined by the surrounding context of the
+inferred expression.  Given a type schema `K` and a record expression `E` of the
+general form `(e0, ..., en, d0 : t0, ..., dm : tm)`, then inference proceeds as
+follows.
+
+If `K` is a record type schema of the form `(K0, ..., Kn, {d0 : P0, ...., dm :
+Pm})`, then:
+  - Each `ei` is inferred with context type schema `Ki` to have type `Ti`
+  - Each `ti` is inferred with context type schema `Pi` to have type `Si`
+  - The type of `E` is `(T0, ..., Tn, {d0 : S0, ...., dm : Sm})`
+
+If `K` is any other type schema:
+  - Each `ei` is inferred with context type schema `_` to have type `Ti`
+  - Each `ti` is inferred with context type schema `_` to have type `Si`
+  - The type of `E` is `(T0, ..., Tn, {d0 : S0, ...., dm : Sm})`
+
+Note that contrary to the practice for runtime checked covariant nominal types,
+we do not prefer the context type over the more precise upwards type.  The
+following example illustrates this:
+
+```dart
+  // No error, inferred type of the record is (int, double)
+  (num, num) r = (3, 3.5)..$0.isEven;
+```
 
-Type inference and promotion flows through records in much the same way it does
-for instances of generic classes (which are covariant in Dart just like record
-fields are) and collection literals.
+Note that implicit casts are not considered to be applied as part of inference,
+hence:
+
+```dart
+  // Static error, inferred type of the record is (dynamic, double)
+  (num, num) r = (3 as dynamic, 3.5);
+```
```

Comment:
OK, that's a natural consequence of using bottom-up inference: The resulting type is `(dynamic, double)`, so it isn't `dynamic`, so there's no implicit cast. Yet ...

Do we actually want to maintain that a cast from dynamic must be explicit in this particular situation? That feels like a surprising deviation from the approach otherwise taken in Dart (that is: by default, dynamic expressions get a cast implicitly; by lint or by `--something-strict`, developers can choose to handle every occurrence of the type dynamic explicitly).

This also means that we cannot apply the perspective that record-to-record assignability is treated in a way which is essentially field-by-field (and that perspective is otherwise applicable quite often).

Line 360-361 actually says that the context type schema is propagated down into record literals (similarly to the propagation of context type schemas into any other composite, e.g., list literals), so we could say that inference transforms an expression `e` with static type `dynamic` into `e as T`, where `T` is obtained from the context type schema in the same way as in those other cases:

```dart
X f<X>(X x) => x;

class C<X> {
  final X x;
  C(this.x);
}

void main() {
  var d = 1 as dynamic;
  int i = d; // Implicit cast, OK.
  int j = f(d); // Implicit cast, OK.
  C<int> c = C(d); // Implicit cast, OK.
  List<int> list = [d]; // Implicit cast, OK.
  Set<int> set = {d}; // Implicit cast, OK.
  Map<int, int> map = {d: d}; // Implicit casts, OK.
  (int,) r = (d,); // Error or implicit cast?
}
```

I believe this approach would also work if we wish to change type inference to be bottom-up with respect to soundly co/contra-variant type variables. The situation (and the surprise) would be the same:

```dart
class C<out X, inout Y> {
  final X x;
  Y y;
  C(this.x, this.y);
}

void main() {
  var d = 1 as dynamic;
  C<int, int> c = C(d, d); // Error or implicit casts?
}
```

Reply:
Maybe another minor argument in defence of casting from `dynamic` is the potential spreading of records into argument lists. In the following example the spreading will not work, forcing the manual spreading. The manual spreading may appear especially surprising in this case because it will work without the explicit downcast from `dynamic`, making the reader questioning why the spreading was done manually in the first place.

```dart
foo(Funcion(int, String) f, (dynamic, String) a) {
  // This will not work.
  f(...a); // '...a' is a hypothetical syntax for spreading arguments.

  // This will work.
  f(a.$0, a.$1); // The downcast from 'dynamic' will be inserted here.
}
```

Reply:
> OK, that's a natural consequence of using bottom-up inference: The resulting type is (dynamic, double), so it isn't dynamic, so there's no implicit cast. Yet ...

Github PR comment locations aren't precise enough for me to know what you are referring to by "that".

> Do we actually want to maintain that a cast from dynamic must be explicit in this particular situation? That feels like a surprising deviation from the approach otherwise taken in Dart (that is: by default, dynamic expressions get a cast implicitly; by lint or by --something-strict, developers can choose to handle every occurrence of the type dynamic explicitly).

The approach I have taken here is consistent with what we currently do elsewhere.  The alternatives are a deviation.  I expand on this [here](https://github.com/dart-lang/language/issues/2488), let's continue discussion there.

Reply:
Just for the record: "that" is the error in line 382.